src/wallet/rpcdump.cpp

```diff
@@ -854,211 +941,162 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+        // Force users to provide the witness script in its field rather than redeemscript
+        if (!strRedeemScript.empty() && script.IsPayToWitnessScriptHash()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "P2WSH addresses have an empty redeemscript. Please provide the witnessscript instead.");
+        }
 
-                if (::IsMine(*pwallet, scriptRawPubKey) == ISMINE_SPENDABLE) {
-                    throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this address or script");
+        // Verify and process input data
+        bool have_solving_data = import_data.redeemscript || import_data.witnessscript || pubkeys.size() || privkeys.size();
+        if (have_solving_data) {
+            // Match up data in import_data with the scriptPubKey in script.
+            auto error = RecurseImportData(script, import_data, ScriptContext::TOP);
+
+            // Check that all required keys for solvability are provided.
+            if (error.empty()) {
+                for (const auto& require_key : import_data.require_keys) {
+                    if (pubkeys.count(require_key) == 0 && privkeys.count(require_key) == 0) {
+                        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Key does not match address destination");
+                    }
                 }
+            }
 
-                pwallet->MarkDirty();
+            // Verify whether the watchonly option corresponds to the availability of private keys.
+            if (!watchOnly && std::any_of(import_data.used_keys.begin(), import_data.used_keys.end(), [&](const CKeyID& used_key){ return privkeys.count(used_key) == 0; })) {
+                warnings.push_back("Assuming watchonly as not all private keys are provided.");
+            }
 
-                if (!pwallet->AddWatchOnly(scriptRawPubKey, timestamp)) {
-                    throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
+            if (!error.empty()) {
+                warnings.push_back("Importing as non-solvable: " + error + ". If this is intentional, don't provide any keys, pubkeys, witnessscript, or redeemscript.");
+                import_data = ImportData();
+                have_solving_data = false;
+            } else {
+                // RecurseImportData() removes any relevant redeemscript/witnessscript from import_data, so we can use that to discover if a superfluous one was provided.
+                if (import_data.redeemscript) warnings.push_back("Ignoring redeemscript as this is not a P2SH script.");
+                if (import_data.witnessscript) warnings.push_back("Ignoring witnessscript as this is not a (P2SH-)P2WSH script.");
+                for (const auto& privkey : privkeys) {
+                    if (import_data.used_keys.count(privkey.first) == 0) {
+                        warnings.push_back("Ignoring irrelevant private key.");
+                        break;
```

Comment:
The warning says that the irrelevant private key will be ignored, but it looks like the private key will still be imported anyways.

Also, why exit this loop when an irrelevant private key is encountered?

Reply:
Nice catch, fixed.